src/pip/_internal/network/download.py

```diff
@@ -185,26 +217,77 @@ def __init__(
         self._session = session
         self._progress_bar = progress_bar
 
+        self._queue = Queue()  # type: Queue[bytes]
+        self._chunks = iter(())  # type: Iterator[bytes]
+        self._count = 0  # number of downloading files
+        self._lock = Lock()
+
+    def _files_to_download(self, links, location):
+        # type: (Iterable[Link], str) -> Iterable[_FileToDownload]
+        file_to_download = partial(_FileToDownload, location, self._session)
+        for link in links:
+            file = file_to_download(link)
+            _log_response(file.size, file.is_cached, link)
+            yield file
+
+    @property
+    def _is_downloading(self):
+        # type: () -> bool
+        with self._lock:
+            return self._count > 0
+
+    def _iter_chunks(self):
+        # type: () -> Iterable[bytes]
+        while self._is_downloading:
+            yield self._queue.get()
+
+    def _display_progress(self, files):
+        # type: (List[_FileToDownload]) -> None
+        total_size = sum(file.size for file in files)
+        if not _should_hide_progress(self._progress_bar, False, total_size):
+            assert not self._is_downloading
+            self._count = len(files)
+            progress = DownloadProgressProvider(self._progress_bar, total_size)
+            self._chunks = progress(self._iter_chunks())
+
+    def _update_progress_bar(self, chunk):
+        # type: (bytes) -> None
+        """Work around the dead lock (?) around next(self._chunks)."""
+        self._queue.put(chunk)
+        while True:
+            # Calling any of the generator methods when the generator
+            # is already executing raises a ValueError:
+            # https://docs.python.org/dev/reference/expressions.html
+            try:
+                next(self._chunks)
+            except ValueError:
+                continue
+            else:
+                break
```

Comment:
I figured that this mutex re-implementation is what is causing the slow down, e.g. using `sleep(0.1)` in place of `continue` would bring the (estimated) download speed back to my bandwidth.

Reply:
Fixed by using a separate lock.